zipkin-server/src/test/java/zipkin2/server/internal/eureka/ITZipkinEureka.java

```diff
@@ -0,0 +1,138 @@
+import org.springframework.test.context.DynamicPropertyRegistry;
+import org.springframework.test.context.DynamicPropertySource;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import zipkin.server.ZipkinServer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.testcontainers.utility.DockerImageName.parse;
+
+/**
+ * Integration test for Eureka, which validates with <a href="https://github.com/Netflix/eureka/wiki/Eureka-REST-operations">Eureka REST operations</a>
+ */
+@SpringBootTest(
+  classes = ZipkinServer.class,
+  webEnvironment = SpringBootTest.WebEnvironment.NONE, // RANDOM_PORT requires spring-web
+  properties = {
+    "server.port=0",
+    "spring.config.name=zipkin-server",
+    "spring.main.banner-mode=off",
+    "zipkin.storage.type=", // cheat and test empty storage type
+    "zipkin.collector.http.enabled=false",
+    "zipkin.query.enabled=false",
+    "zipkin.ui.enabled=false"
+  })
+@Testcontainers
+@TestMethodOrder(OrderAnnotation.class) // so that deregistration tests don't flake the others.
+class ITZipkinEureka {
+
+  @Container static EurekaContainer eureka = new EurekaContainer();
+
+  @Autowired Server zipkin;
+  OkHttpClient client = new OkHttpClient.Builder().followRedirects(false).build();
+
+  /** Get the serviceUrl of the Eureka container prior to booting Zipkin. */
+  @DynamicPropertySource static void propertyOverride(DynamicPropertyRegistry registry) {
+    registry.add("zipkin.discovery.eureka.serviceUrl", eureka::serviceUrl);
+  }
+
+  @Test @Order(1) void registersInEureka() throws Exception {
+    // Notice the app name always coerces to uppercase
+    String json = getEurekaAsString("/apps/ZIPKIN");
+
+    // Make sure the health status is OK
+    assertThat(readString(json, "$.application.instance[0].status"))
+      .isEqualTo("UP");
+
+    String zipkinHostname = zipkin.defaultHostname();
+    int zipkinPort = zipkin.activePort().localAddress().getPort();
+
+    // Note: Netflix/Eureka says use hostname, which can conflict on laptops.
+    // Armeria adopts the spring-cloud-netflix convention shown here.
+    assertThat(readString(json, "$.application.instance[0].instanceId"))
+      .isEqualTo(zipkinHostname + ":zipkin:" + zipkinPort);
+
+    // Make sure the vip address is relevant
+    assertThat(readString(json, "$.application.instance[0].vipAddress"))
+      .isEqualTo(zipkinHostname + ":" + zipkinPort);
+  }
+
+  @Test @Order(2) void deregistersOnClose() {
+    zipkin.close();
+    await().untilAsserted( // wait for deregistration
+      () -> assertThat(getEurekaAsString("/apps"))
+        .isEqualTo(
+          "{\"applications\":{\"versions__delta\":\"1\",\"apps__hashcode\":\"\",\"application\":[]}}"));
```

Comment:
Not sure if it could be worth doing `assertThat(readList(getEurekaAsString("/apps"), "$.applications.application")).isEmpty()`. A couple of these look like implementation details and maybe change with eureka versions. Or maybe `getEureka("/apps/ZIPKIN")` would 404.

Reply:
sounds better, thanks!